File: 13_gaussain_libor_market_model_lmm/cpp/Diagnostics.cpp
#include "Diagnostics.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>

void printForwardCurve(
    const GaussianLMM& model
){

    std::cout
        << "\nInitial Forward Curve"
        << std::endl;

    std::cout
        << "============================================================"
        << std::endl;

    std::cout
        << std::setw(8)
        << "Index"
        << std::setw(14)
        << "Reset"
        << std::setw(14)
        << "Payment"
        << std::setw(18)
        << "Forward"
        << std::endl;

    for(int i = 0; i < model.numberOfForwards(); ++i){

        std::cout
            << std::setw(8)
            << i
            << std::setw(14)
            << model.tenor().times()[i]
            << std::setw(14)
            << model.tenor().times()[i + 1]
            << std::setw(18)
            << model.initialForwards()[i]
            << std::endl;
    }
}

void printCorrelationSummary(
    const Matrix& correlationMatrix
){

    double minimum =
        std::numeric_limits<double>::max();

    double maximum =
        -std::numeric_limits<double>::max();

    for(const auto& row : correlationMatrix){

        for(double value : row){

            minimum =
                std::min(
                    minimum,
                    value
                );

            maximum =
                std::max(
                    maximum,
                    value
                );
        }
    }

    std::cout
        << "\nCorrelation Summary"
        << std::endl;

    std::cout
        << "============================================================"
        << std::endl;

    std::cout
        << "Matrix size : "
        << correlationMatrix.size()
        << " x "
        << correlationMatrix.size()
        << std::endl;

    std::cout
        << "Minimum     : "
        << minimum
        << std::endl;

    std::cout
        << "Maximum     : "
        << maximum
        << std::endl;
}

void printSimulationSummary(
    const GaussianLMM& model,
    const LMMSimulationResult& simulation
){

    std::cout
        << "\nSimulation Summary"
        << std::endl;

    std::cout
        << "============================================================"
        << std::endl;

    std::cout
        << "Paths       : "
        << simulation.forwardPaths.size()
        << std::endl;

    std::cout
        << "Time points : "
        << simulation.times.size()
        << std::endl;

    std::cout
        << "Forwards    : "
        << model.numberOfForwards()
        << std::endl;

    std::cout
        << "Factors     : "
        << model.numberOfFactors()
        << std::endl;

    std::cout
        << "Terminal DF : "
        << model.initialTerminalDiscountFactor()
        << std::endl;
}

void printPricingSummary(
    double capletPrice,
    double swaptionPrice
){

    std::cout
        << "\nGaussian LMM Pricing Summary"
        << std::endl;

    std::cout
        << "============================================================"
        << std::endl;

    std::cout
        << "Caplet price          : "
        << capletPrice
        << std::endl;

    std::cout
        << "Payer swaption price : "
        << swaptionPrice
        << std::endl;
}

File: 13_gaussain_libor_market_model_lmm/python/diagnostics.py
# ✝ In memory of my beloved grandfather - Heshun Feng (冯和顺), who passed away today (11th July 2026)
# His love and encouragement will forever inspire me.


"""
diagnostics.py

Diagnostics for the Gaussian LMM.
"""

import numpy as np
import pandas as pd


def forward_curve_table(
    tenor_structure,
    initial_forwards
):
    """
    Return initial forward curve as a DataFrame.
    """

    return pd.DataFrame(
        {
            "reset_time": tenor_structure.times[:-1],
            "payment_time": tenor_structure.times[1:],
            "accrual": tenor_structure.accruals,
            "initial_forward": initial_forwards
        }
    )


def summarize_simulation(
    times,
    forward_paths,
    tenor_structure
):
    """
    Summarize simulated forward rates at their reset dates.
    """

    rows = []

    for i in range(
        tenor_structure.n_forwards
    ):

        reset_time = tenor_structure.times[i]

        simulation_index = int(
            np.argmin(
                np.abs(
                    times
                    -
                    reset_time
                )
            )
        )

        values = forward_paths[
            :,
            simulation_index,
            i
        ]

        rows.append(
            {
                "forward_index": i,
                "reset_time": reset_time,
                "mean": values.mean(),
                "std": values.std(),
                "min": values.min(),
                "max": values.max()
            }
        )

    return pd.DataFrame(rows)


def check_correlation_matrix(
    correlation_matrix
):
    """
    Return basic correlation diagnostics.
    """

    eigenvalues = np.linalg.eigvalsh(
        correlation_matrix
    )

    return {
        "minimum_eigenvalue": float(
            eigenvalues.min()
        ),
        "maximum_eigenvalue": float(
            eigenvalues.max()
        ),
        "symmetric": bool(
            np.allclose(
                correlation_matrix,
                correlation_matrix.T
            )
        ),
        "unit_diagonal": bool(
            np.allclose(
                np.diag(
                    correlation_matrix
                ),
                1.0
            )
        )
    }


def print_pricing_summary(
    caplet_price,
    swaption_price
):
    """
    Print caplet and swaption prices.
    """

    print("Gaussian LMM Pricing Summary")
    print("=" * 70)

    print(
        f"Caplet price          : {caplet_price:,.6f}"
    )

    print(
        f"Payer swaption price : {swaption_price:,.6f}"
    )